src/backend/access/transam/twophase.c

```diff
@@ -2459,3 +2460,85 @@ getTwoPhaseOldestPreparedTransactionXLogRecPtr(prepared_transaction_agg_state *p
 	return oldest;
 
 }  /* end getTwoPhaseOldestPreparedTransactionXLogRecPtr */
+
+bool
+RemovePendingDeletesForPreparedTransactions()
+{
+	HASH_SEQ_STATUS scan_status;
+	prpt_map   *entry;
+	XLogReaderState *xlogreader;
+	volatile bool result = true;
+	XLogRecord *xlogrec = NULL;
+	MemoryContext oldcontext = CurrentMemoryContext;
+
+	if (NULL == crashRecoverPostCheckpointPreparedTransactions_map_ht)
+		return result;
+
+	xlogreader = XLogReaderAllocate(&read_local_xlog_page, NULL);
+	if (!xlogreader)
+		ereport(ERROR,
+				(errcode(ERRCODE_OUT_OF_MEMORY),
+				 errmsg("out of memory"),
+		   errdetail("Failed while allocating an XLog reading processor.")));
+
+	hash_seq_init(&scan_status,
+				  crashRecoverPostCheckpointPreparedTransactions_map_ht);
+	while ((entry = (prpt_map *) hash_seq_search(&scan_status)) != NULL)
+	{
+		char	   *errormsg = NULL;
+		TwoPhaseFileHeader *hdr;
+
+		if (entry->xlogrecptr == InvalidXLogRecPtr)
+			continue;
+
+		int savedInterruptHoldoffCount = InterruptHoldoffCount;
```

Comment:
When sigsetjmp (PG_TRY) returns after siglongjmp, it restores stack only. The values of register variables are unpredictable (https://www.ibm.com/docs/en/zos/3.1.0?topic=functions-sigsetjmp-save-stack-environment-signal-mask). So at least, savedInterruptHoldoffCount should be declared as volatile to avoid its mapping to a register.